### PR TITLE
GSoC 2020 report: André Perez Maselco

### PR DESCRIPTION
**Pull request by andreperezmaselco**
Report describing the work done during Google Summer of Code 2020.

### CODE REVIEW

**Comment by andreperezmaselco** [comment]
@googlebot I signed it!

File: docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+##### Organization:
+##### [Android Graphics Tools Team](https://summerofcode.withgoogle.com/organizations/6102438181863424/)
+
+##### Mentors:
+##### [Alastair Donaldson](https://github.com/afd)
+##### [Paul Thomson](https://github.com/paulthomson)
+
+## Abstract
+
+During the *Google Summer of Code 2020* program, I worked on the development of new semantics-preserving transformations, fuzzer passes and enhancements for the *spirv-fuzz* tool.
```
Comment (afd) [inline docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md:18]: Perhaps you could link to this URL for spirv-fuzz:

https://github.com/google/graphicsfuzz/blob/master/docs/finding-a-vulkan-driver-bug-using-spirv-fuzz.md

File: docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+Also, I developed [Amber Fuzz](https://github.com/andreperezmaselco/amber-fuzz), a simple graphics tool that takes an [Amber](https://github.com/google/amber) script as input and outputs a new script by fuzzing the reference shaders.
+
+## Introduction
+Graphics shaders are programs to run on a Graphics Processing Unit. Shaders can be written using the *OpenGL Shading Language* or *SPIR-V* (a simple binary intermediate language). Since shaders are programs, they need to be compiled into specific GPU instructions. A question that arises is what if the shader compiler has bugs? Some tools have been developed to deal with this problem.
```
Comment (afd) [inline docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md:22]: I would rephrase to say that shaders are written in some shading language, and that the OpenGL shading language and SPIR-V are two examples.  (There are more, e.g. HLSL and MetalSL.)

File: docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+
+[SPIRV-Tools](https://github.com/KhronosGroup/SPIRV-Tools) is a collection of tools to work on *SPIR-V* modules.
+One of these tools is *spirv-fuzz*. It applies semantics-preserving transformations to a *SPIR-V* module. Theoretically, a semantics-preserving sentence should not change the behaviour of the variant code. If it has changed, then an error has occurred halfway, probably a compiler bug. In this way, the reference and variant shaders can be compared to check the differences. This approach, known as fuzzing, is part of technique called metamorphic testing.
```
Comment (afd) [inline docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md:25]: I think "sentence" is a typo.  And the use of the word "variant" comes out of the blue - you should explain what you mean by reference and variant.

File: docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+%12 = OpCompositeConstruct %5 %14 %16</strong>
+      OpReturn
+      OpFunctionEnd
+        </code>
+      </pre>
+    </td>
+  </tr>
+</table>
+
+This project was concerned to use the metamorphic testing approach to extend *spirv-fuzz* by adding new transformations to cover features of *SPIR-V* that are not yet very widely-used, but that will be in the future.
```
Comment (afd) [inline docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md:92]: "was concerned to use" -> "involved using"

File: docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+Also, the project used tools such as [CLSmith](https://github.com/ChrisLidbury/CLSmith) and [clspv](https://github.com/google/clspv) to improve *spirv-fuzz*.
+
+## Development
+
+Transformations and fuzzer passes work together to modify the shader source code.
+The fuzzer pass iterates over the *SPIR-V* module and decides whether a transformation will be applied.
+Below are the transformations, fuzzer passes and improvements developed during the program.
```
Comment (afd) [inline docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md:99]: I suggest using the first person here: "that I developed".

File: docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+
+- [#3412](https://github.com/KhronosGroup/SPIRV-Tools/pull/3412): Add Vector Shuffle Instruction
+
+- [#3439](https://github.com/KhronosGroup/SPIRV-Tools/pull/3439): Add Image Sample Unused Components
+
+- [#3597](https://github.com/KhronosGroup/SPIRV-Tools/pull/3597): Make Vector Operations Dynamic
+
+- [#3517](https://github.com/KhronosGroup/SPIRV-Tools/pull/3517): Inline Function
+
+- Add Bit Instruction Synonym (Work In Progress)
```
Comment (afd) [inline docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md:121]: You can of course do another PR after the deadline to add the link for this if you like.

File: docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+- [#3557](https://github.com/KhronosGroup/SPIRV-Tools/issues/3557): Split bit instructions into most fundamental operations
+
+- [#3588](https://github.com/KhronosGroup/SPIRV-Tools/issues/3588): Dynamically extract and insert vector components
+
+- [#3646](https://github.com/KhronosGroup/SPIRV-Tools/issues/3646): Replace decimal arithmetic with binary arithmetic
+
+## Conclusion
+It was great to participate in the *Google Summer of Code 2020* program. I was able to learn more about computer graphics, improve code writing and project collaboration. Finally, I would like to thank my mentors, Alastair Donaldson and Paul Thomson, for the opportunity.
+
+## References
```
Comment (afd) [inline docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md:196]: It's nice that you have references!

You should refer to the references earlier in the text.  E.g. you could refer to [1] if you mention the glsl-fuzz tool that was the precursor to spirv-fuzz.  (You might then also want to mention our paper "Automated testing for Graphics Shader Compilers" that has Paul as an author.)  I suppose you wanted to reference Abel and Jiradet's reports because you were inspired by their structure; in that case perhaps you should do it in a short note at the end of the conclusion.

In general when writing it's not good style to have items in a bibliography if you don't refer to those items from the main document.

File: docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+      OpReturn
+      OpFunctionEnd
+        </code>
+      </pre>
+    </td>
+  </tr>
+</table>
+
+This project was concerned to use the metamorphic testing approach to extend *spirv-fuzz* by adding new transformations to cover features of *SPIR-V* that are not yet very widely-used, but that will be in the future.
+Also, the project used tools such as [CLSmith](https://github.com/ChrisLidbury/CLSmith) and [clspv](https://github.com/google/clspv) to improve *spirv-fuzz*.
```
Comment (afd) [inline docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md:93]: Somewhere in the report it would be nice if you could say a little more about this.

**Review by afd — COMMENTED** [review]
Good report @andreperezmaselco!

I have made some comments.

How about you address them - plus any comments that @paulthomson may have - and then we can merge?

File: docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md (hunk 3 above)
Comment (paulthomson) [reply docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md:25]: I agree with Ally. Also, I would remove the word "halfway". 


Ideas:

It applies semantics-preserving transformations to a *SPIR-V* module, that we call the _reference_ module, to obtain a modified _variant_ module.


should not change the behaviour of the code

In this way, the reference and variant modules can be run, and the outputs compared. The output is usually an image. If the images differ, this probably indicates a shader compiler a bug. 


This approach, known as fuzzing, is part of technique called metamorphic testing.
->
This is a specialised form of _fuzzing_ that uses a technique called _metamorphic testing_.


Consider making _metamorphic testing_ a link to our paper or another paper, or use this approach to link to a reference at the end: https://stackoverflow.com/questions/26587527/cite-a-paper-using-github-markdown-syntax

File: docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+
+[SPIRV-Tools](https://github.com/KhronosGroup/SPIRV-Tools) is a collection of tools to work on *SPIR-V* modules.
+One of these tools is *spirv-fuzz*. It applies semantics-preserving transformations to a *SPIR-V* module. Theoretically, a semantics-preserving sentence should not change the behaviour of the variant code. If it has changed, then an error has occurred halfway, probably a compiler bug. In this way, the reference and variant shaders can be compared to check the differences. This approach, known as fuzzing, is part of technique called metamorphic testing. 
+
+The following is an example of transformation that replaces an `OpVectorTimesScalar` instruction with its mathematical definition. The vector components are extracted and multiplied by the scalar. Finally, the new vector is constructed using the multiplied components.
```
Comment (paulthomson) [inline docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md:27]: As discussed, you could start a new section here. At least make sure you say something like "The following is an example of transformation that I implemented" or "that I proposed and implemented" (if true).

File: docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md (hunk 4 above)
Comment (paulthomson) [inline docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md:92]: I would also remove: "to cover features of *SPIR-V* that are not yet very widely-used, but that will be in the future"

I don't think this is true. We want to target features that are used and will continue to be used. You could say that you added transformations to cover instructions and patterns of SPIR-V that were not previously covered by existing spirv-fuzz transformations.

Comment (afd) [reply docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md:92]: As a counter-argument to what @paulthomson said: we have been keen to cover parts of SPIR-V that are not widely-used by GLSL shaders that get converted to SPIR-V by glslang.  I think some of the transformations that we make lead to SPIR-V that you could never get from one of the standard front-ends, so in that sense it's SPIR-V that isn't widely-used.  You might want to discuss that a bit Andre.

File: docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md (hunk 1 above)
Comment (andreperezmaselco) [reply docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md:18]: Done.

File: docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md (hunk 2 above)
Comment (andreperezmaselco) [reply docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md:22]: Done.

File: docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md (hunk 3 above)
Comment (andreperezmaselco) [reply docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md:25]: Done.

File: docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md (hunk 4 above)
Comment (andreperezmaselco) [reply docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md:92]: Done.

File: docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md (hunk 5 above)
Comment (andreperezmaselco) [reply docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md:99]: Done.

File: docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md (hunk 9 above)
Comment (andreperezmaselco) [reply docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md:27]: Done.

File: docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md (hunk 8 above)
Comment (andreperezmaselco) [reply docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md:93]: Done.

File: docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md (hunk 7 above)
Comment (andreperezmaselco) [reply docs/summer-of-code-ideas/summer-of-code-reports/gsoc-2020-andreperezmaselco.md:196]: About the reports of Abel and Jiradet, I would like to leave them as references since I got in touch with the concepts I learned via their reports first.

**Review by paulthomson — APPROVED** [review]
Thanks for doing a great GSoC project!
